### ISSUE TITLE
respect return type of `__new__`

### ISSUE THREAD

**Issue by nickdrozd** [astral-sh/ty#281]
### Summary

This code passes Ty without any warnings. But when run it will fail half the time.

```python
from __future__ import annotations

from random import randint
from typing import TYPE_CHECKING

class A:
    def __new__(cls) -> int | A:
        if randint(0, 1):
            return object.__new__(cls)
        else:
            return randint(0, 100)

def f(x: A) -> None:
    assert isinstance(x, A)

a = A()

f(a)
```

`f` expects value of type `A`. But `A.__new__` might return `int`. So constructor of `A` should not be assumed to always return `A`. Ty should flag call `f(a)`.

### Version

ty 0.0.0-alpha.7

**Comment by carljm** [comment]
Thanks for the report! Yes, it's a known issue that we don't yet handle `__new__` return type properly, we currently assume it returns an instance of the class. But I don't think we had an issue open for it yet, that I can find, so will keep this one!

**Comment by dhruvmanila** [comment]
This also affects calls like `slice(a, b, c)` where all parameters are of type `int` as the reveal type is `slice[Any, Any, Any]`. The revealed type of the `__new__` call (which is an overloaded function) is `slice[int, int, int]`.
